Unconstrained type parameters and `??` (#1323)
dotnet/csharpstandard: standard/expressions.md

The type of the expression `a ?? b` depends on which implicit conversions are available on the operands. In order of preference, the type of `a ?? b` is `A₀`, `A`, or `B`, where `A` is the type of `a` (provided that `a` has a type), `B` is the type of `b`(provided that `b` has a type), and `A₀` is the underlying type of `A` if `A` is a nullable value type, or `A` otherwise. Specifically, `a ?? b` is processed as follows:

> - If `A` exists and is an unmanaged type (§8.8) or known to be a non-nullable value type, a compile-time error occurs.

Reviewer: Does "non-nullable value type" not subsume "unmanaged type"?

Reviewer: Ah, I missed the pointers conversation.

- Otherwise, if `A` exists and an implicit conversion exists from `b` to `A`, the result type is `A`. At run-time, `a` is first evaluated. If `a` is not `null`, `a` becomes the result. Otherwise, `b` is evaluated and converted to type `A`, and this becomes the result.
- Otherwise, if `A` exists and is a nullable value type, `b` has a type `B` and an implicit conversion exists from `A₀` to `B`, the result type is `B`. At run-time, `a` is first evaluated. If `a` is not `null`, `a` is unwrapped to type `A₀` and converted to type `B`, and this becomes the result. Otherwise, `b` is evaluated and becomes the result.
- Otherwise, if `b` has a type `B` and an implicit conversion exists from `a` to `B`, the result type is `B`. At run-time, `a` is first evaluated. If `a` is not `null`, `a` is converted to type `B`, and this becomes the result. Otherwise, `b` is evaluated and becomes the result.
- Otherwise, `a` and `b` are incompatible, and a compile-time error occurs.

> *Example*:
>
> <!-- Example: {template:"standalone-console-without-using", name:"NullCoalescingGenerics"} -->
> ```csharp
> T M<T>(T a, T b) => a ?? b;
>
> string s = M(null, "text");
> int i = M(19, 23);
> ```
>

> > The parameters to `M` are unconstrained. Therefore, the type argument can be a reference type, or nullable value type, as shown in the first call to `M`. The type argument can also be a non-nullable value type, as shown in the second call to `M`. When the type argument is a non-nullable value type, the expression `a ?? b` is `a`.

Reviewer 2: ```suggestion
> The type parameter `T` for method `M` is unconstrained. Therefore, the type argument can be a reference type, or nullable value type, as shown in the first call to `M`. The type argument can also be a non-nullable value type, as shown in the second call to `M`. When the type argument is a non-nullable value type, the expression `a ?? b` is `a`.
```

> > The type parameter `T` for method `M` is unconstrained. Therefore, the type argument can be a reference type, or nullable value type, as shown in the first call to `M`. The type argument can also be a non-nullable value type, as shown in the second call to `M`. When the type argument is a non-nullable value type, the expression `a ?? b` is `a`.

Reviewer 2: ```suggestion
> The type parameter `T` for method `M` is unconstrained. Therefore, the type argument can be a reference type, or nullable value type, as shown in the first call to `M`. The type argument can also be a non-nullable value type, as shown in the second call to `M`. When the type argument is a non-nullable value type, the value of the expression `a ?? b` is `a`.
```

The type of the expression `a ?? b` depends on which implicit conversions are available on the operands. In order of preference, the type of `a ?? b` is `A₀`, `A`, or `B`, where `A` is the type of `a` (provided that `a` has a type), `B` is the type of `b`(provided that `b` has a type), and `A₀` is the underlying type of `A` if `A` is a nullable value type, or `A` otherwise. Specifically, `a ?? b` is processed as follows:

> - If `A` exists and is not a nullable value type, a compile-time error occurs.

Reviewer 3: I’ll review this properly later but for now did you mean:

```suggestion
- If `A` exists and is a non-nullable value type a compile-time error occurs.
```

as the current wording would *exclude* reference types? Should other types be invalid, e.g. pointer types?

Reviewer 2: I share the same confusion as Nigel here. And I don't know offhand whether a type parameter that is constrained (e.g. `T where T : struct`) counts as "a non-nullable value type" or just "known to be a non-nullable value type".

Author: Does `T where T : struct` counts as a non-nullable value type.  I think so. Here's a small test:

```csharp
int? number = null;
int? number2 = 4;
IsNullableValueTypeAStruct(number); // Error: CS0453 "The type 'int?' must be a non-nullable value type in order to use it as a parameter 'T' in the generic type or method 'NullCoaelescing.IsNullableValueTypeAStruct<T>(T)'"
NullCoalescingStruct(5, 12);
NullCoalescingStruct(number, number2); // // Error: CS0453

void IsNullableValueTypeAStruct<T>(T a) where T : struct  { }

void NullCoalescingStruct<T>(T a, T b) where T : struct 
{
    T result = a ?? b; // Error: CS0019 "Operator '??' cannot be applied to operands of type 'T' and 'T'"
}
```

Author: I applied Nigel's suggestion and excluded unmanaged types (which includes pointer types). There is some duplication, but it feels more correct.

I unresolved the conversation because of Jon's question.

Reviewer 4: I concur Bill; if I recall correctly the intention of the C# 2.0 design was that `where T:struct` definitely ensures that `T` is a non-nullable value type.

Otherwise, `a` and `b` are incompatible, and a compile-time error occurs.

> *Example*:
>
> <!-- Example: {template:"standalone-console-without-using", name:"DeclarationExpressions1"} -->
> ```csharp
> T M<T>(T a, T b) => a ?? b;
>
> string s = M(null, "text");
> int i = M(19, 23);
> ```
>

> > The parameters to `M` are unconstrained. Therefore, the type arguments can be reference types, or nullable value types, as shown in the first call to `M`. The type arguments can also be non-nullable value types, as shown in the second call to `M`. When the type argument is a non-nullable value type, the expression `a ?? b` is `a`.

Reviewer 2: ```suggestion
> The parameters to `M` are unconstrained. Therefore, the type argument can be a reference type, or nullable value type, as shown in the first call to `M`. The type argument can also be a non-nullable value type, as shown in the second call to `M`. When the type argument is a non-nullable value type, the expression `a ?? b` is `a`.
```

(Just because there's only a single type parameter, and therefore a single type argument.)

Reviewer 2: For "the expression `a ?? b` is `a`" should we have "is evaluated equivalently to" or something similar?

Reviewer 3: @jskeet – No, if `a` is a non-nullable value type then `a` is non-null (think of the test as `(object)a == null` which for an IL target is a `box` and test) and so `a` is the result – it is not different handling *per se*. What is needed is to revise the bulleted list to cover this, see my earlier comment.

Reviewer 2: My point is that the "is" here is a bit ambiguous. Perhaps "the value of the expression `a ?? b` is the value of `a`" or something like that? Just "is" seems wrong to me.

Reviewer 3: @jskeet – I see, your "is evaluated equivalently to" threw me and “is the value of” is a lot better. But that leaves us with the problem that this example is not supported by the text, we could say its aspirational! ;-)

>

> > *end example*

Reviewer 3: I don’t think this is there yet.

Bullets 2 thru 5 are all of the form “Otherwise, if `A` exists & <cond>… At runtime…”. This is a compile-time condition followed by the runtime actions. If `A` is an unconstrained type parameter then at compile-time it certainly exists (a type-parameter cannot be bound to a non-type), but what it is is not known until runtime. So this gets you to bullet 6, which is also where you end up if A does not exist (i.e. `a` is `null`), and this is incorrect for the dynamic case – it is correct for `null`.

There needs to be a bullet 5A (i.e. an inserted one), which covers runtime resolution and may be defined either from scratch, may include applying 2-5 at runtime, or whatever is needed.

This example can then refer to the rules and say which result is used in each case.

BTW I think line 4832 should be bullet 7.

Sorry, no time to attempt a starting point for this right now.

Author: I did add line 4832 to the bullet list.

The other concerns are ones I'd like to address in a new PR. The condition Nigel describes was already in the standard. A different task for this part makes sense to me.
